### ISSUE TITLE
PostgreSql Question

### ISSUE THREAD

**Issue by murst** [blogifierdotnet/Blogifier#30]
I'm trying to switch over from using Azure to using Digital Ocean. I want to switch from SqlServer to PostgreSql, but I'm not sure what exactly is needed. I'm not getting the code from Nuget - I'm getting it directly from github so that I can modify stuff.

I have a version of linux running, and .net core is running as well. I'm just not clear on how to connect to the database.

Is the code for Blogifier.Core.PostgreSql available somewhere? Can I just use the Blogifier.Core project and point my webapp to PostgreSql for the database? I guess I'm not exactly clear on how to make this work on Linux with a different type of database.

**Comment by rxtur** [comment]
The PostgreSql package built on the fly from Core by replacing SQL Server specific code with PostgreSql code, there is no separate project.

If you asking about connection string, this is what I'm using on Digital Ocean:

`"User ID=blogifier;Password=yourpass;Host=localhost;Port=5432;Database=blogifier;Pooling=true;"`

I did have to do some odd things like have to name user and DB with the same name and run SQL script to create database because EF migrations would not run, but I kind of expected some extra effort from Linux environment.

I really need to write a blog post on setting up with Digital Ocean, will do when have time.

**Comment by murst** [comment]
I have a few improvements you guys might be interested in.

I modified the Blogifier.Core project so that it doesn't have any dependencies on SqlServer (or Postgres, etc). You now pass the database options from the host application to Blogifier.Core, and it configures itself for the appropriate database. This should allow for a single Blogifier.Core project that supports many databases, not just SqlServer. Also, i don't think there's a need anymore for a separate Blogifier.Core.PostgreSql package if you guys accept these changes.

Also, I added some attributes to BaseEntity, which allowed me to do a migration to PostgreSql. This still needs to be tested for a clean install of SqlServer (I didn't do that), but I was able to create a blog without running any SQL scripts on PostgreSql, and everything seems to be working properly.

The commit with all the necessary changes is located at: https://github.com/murst/Blogifier.Core/commit/12385cb8deb59b8e4f79867c24655655371c159f

There's also a small fix to retrieving a post, because I ran into an error when my blog did not contain any posts.

**Comment by murst** [comment]
Edit: Also, I just realized that my startup.cs has some code about forwardedheaders. That was needed to get it to work on linux. Just ignore that. Sorry.

**Comment by murst** [comment]
Just a quick update: the changes I made to BaseEntity seemed to work on my Windows PostgreSql db; however, it doesn't appear to work on Linux.  Basically, I can't get the primary keys for the tables to auto-increment in PostgreSql on linux using code-first. I'll try working on it some more later.

**Comment by rxtur** [comment]
Ok, please let know if you'll get it working - I would love to consolidate all in single package.

**Comment by murst** [comment]
You can actually use the code in the commit to make everything into a single package already. What's not working is the database migrations in PostgresSql on Linux, but AFAIK, that doesn't work in the current packages either. 

I'll let you know if I come up with a fix for the primary key in BaseEntity.

**Comment by murst** [comment]
ok, got it working now. Had to update the migrations since they had specific code for SqlServer, but nothing for PostgreSql.

https://github.com/murst/Blogifier.Core/commit/ca2d7c359e0734ec8d3bea92506ca81a3a2899d5

If you update the code w/ the changes from the 2 commits linked in this thread, you should be able to have a single package that supports both SqlServer and PostgreSql, and the migrations should be working for both. 

If you want to support more databases (such as MySQL), the only thing that would be needed (I think) is just updating the migrations with annotations specific to the new database.

**Comment by rxtur** [comment]
Cool, will check it out tomorrow. That latest commit with migration - it was generated by EF based on annotation you added to base entity, right? Or updated manually?

**Comment by murst** [comment]
I added that manually.

When I added the the attributes to BaseEntity (and before I manually added the annotations), I was able to generate the correct SQL when running it on PostgreSql on Windows. However, it didn't generate the correct SQL on Linux. I'm not sure if that is because of the different OSes, or if it was because I applied the migrations to different states of the database.

After I manually added the annotations to the migration script, the website generated the correct SQL when creating the tables in PostgreSql on Linux (I had already created an empty database in PostgreSql that had the same name as the user in my connection string).
